homeassistant/components/mqtt/__init__.py

```diff
@@ -53,25 +54,62 @@
 ATTR_TOPIC = 'topic'
 ATTR_PAYLOAD = 'payload'
 ATTR_PAYLOAD_TEMPLATE = 'payload_template'
-ATTR_QOS = 'qos'
+ATTR_QOS = CONF_QOS
 ATTR_RETAIN = 'retain'
 
 MAX_RECONNECT_WAIT = 300  # seconds
 
+_VALID_QOS_SCHEMA = vol.All(vol.Coerce(int), vol.In([0, 1, 2]))
+_HBMQTT_CONFIG_SCHEMA = vol.Schema(dict)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_CLIENT_ID): cv.string,
+        vol.Optional(CONF_KEEPALIVE, default=DEFAULT_KEEPALIVE):
+            vol.All(vol.Coerce(int), vol.Range(min=15)),
+        vol.Optional(CONF_BROKER): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT):
+            vol.All(vol.Coerce(int), vol.Range(min=1, max=65535)),
+        vol.Optional(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_CERTIFICATE): vol.IsFile,
+        vol.Optional(CONF_PROTOCOL, default=DEFAULT_PROTOCOL):
+            [PROTOCOL_31, PROTOCOL_311],
+        vol.Optional(CONF_EMBEDDED): _HBMQTT_CONFIG_SCHEMA,
+    }),
+})
+
+MQTT_BASE_PLATFORM_SCHEMA = vol.Schema({
+    vol.Required(CONF_PLATFORM): DOMAIN,
+    vol.Optional(CONF_SCAN_INTERVAL):
+        vol.All(vol.Coerce(int), vol.Range(min=1)),
+    vol.Optional(CONF_QOS, default=DEFAULT_QOS): _VALID_QOS_SCHEMA,
+})
+
 
 # Service call validation schema
-def mqtt_topic(value):
-    """Validate that we can publish using this MQTT topic."""
-    if isinstance(value, str) and all(c not in value for c in '#+\0'):
+def _mqtt_topic(value, invalid_chars='\0'):
+    """Validate MQTT topic."""
+    if isinstance(value, str) and all(c not in value for c in invalid_chars):
         return vol.Length(min=1, max=65535)(value)
     raise vol.Invalid('Invalid MQTT topic name')
 
+
+def valid_publish_topic(value):
+    """Validate that we can publish using this MQTT topic."""
+    return _mqtt_topic(value, invalid_chars='#+\0')
+
+
+def valid_subscribe_topic(value):
```

Comment:
As nothing else uses _mqtt_topic, I will just rename that to valid_subscribe_topic.